crck.1
.\" -*- nroff -*-
.\"
.\" crck 2.1 -- Calculate 16-bit CRC of files.
.\" Copyright (C) 2021 Glitch Works, LLC
.\" http://www.glitchwrks.com/
.\"
.\" Original code Copyright Ben Goldfarb, Computer Development, Inc.
.\"
.\" The following GNU GPL v3 license applies to all Glitch Works, LLC
.\" changes and contributions:
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\"
.\" crck.1 - man page
.\"
.TH crck 1 "2021-05-29" "Version 2.1" "User Commands"

.SH NAME
crck \- calculate CRC value of files

.SH SYNOPSIS
.PD 0
.B crck
.RB [ -t|u|c|i ]
.RI [ filename 
.cc @
... ]
@cc .

.SH DESCRIPTION
Crck calculates and prints a 16-bit CRC value and either a CP/M 128-byte 
sector count or the byte count for the named files.  It is typically used to 
validate files transferred between Unix and CP/M type operating systems. 
Crck uses a CRC compatible with the CP/M and CP/M-86 versions of crck prior to
5.0, as well as the crck command in 8 bit versions of YAM (Yet Another Modem 
program).

.PP
When comparing copies of a text file on Unix and CP/M or similar systems, 
account must be taken of the format of the file on the Unix system to 
determine which option to use to check the file. The file may be in CP/M 
format, with carriage return/line feed line terminators and padding with
control-Z to make the length divisible by 128 (-u option); or it may be in 
Unix format, with newline line delimiters and no padding (-t option); or in 
ITS-binary format (-i option).

.PP
Since CP/M binary files are not 'translated' to Unix, these may be checked 
directly. Note, though, that any binary file to be sent to a CP/M system must
have a length of a multiple of 128 bytes to be checked correctly on both ends.
If it does not, the CP/M system will 'fill' the rest of the last 128-byte 
record, and the CRC checks will be different, even though the data transferred
correctly.

.SH OPTIONS
.TP
.B -t
Perform CRCK on a CP/M text file that is stored as a Unix file. CP/M record 
delimiters (CR/LF) are substituted for Unix newlines and the file is padded to
a 128-byte sector boundary with ^Z (0x1A).

.TP
.B -u
Perform CRCK on a regular Unix file.  Takes whatever is there. This is the 
default option.

.TP
.B -c
Perform CRCK on a CP/M non-text file (such as .COM or .LBR). Handled the same 
as the -u flag except 128-byte sectors are reported rather than bytes.

.TP
.B -i
Perform CRCK on a CP/M non-text file stored in ITS binary format. Ignores the
first four bytes of the file which contain an ITS code.

.TP
Wildcards may, of course, be used and the program may be used in pipelines.

.SH DIAGNOSTICS
Prints proper usage if an incorrect flag is given.

.SH BUGS
Has not been rigorously tested.  May not produce correct results with all 
processor types.  This CRC detects all single bit errors and most, but not 
all, multiple bit errors. This CRC does not detect all two bit errors, unlike
the CRC used with standard synchronous communications protocols.

.SH WARNING
Files with different data may still produce the same crck value.

.SH AUTHORS
Ben Goldfarb
.br
Computer Development, Inc.

.PP
Cleanup for modern UNIX/C by Glitch Works, LLC
.br
http://www.glitchwrks.com/


crck.c
/*
 *    crck.c  --  Unix version of the CP/M utility
 *
 *    Usage:
 *        crck [-t|c|u|i] [filename ... ]
 *
 *    Calculates a 16-bit cyclic redundancy check on its input files and
 *    prints the resulting number in hexadecimal.  CRCK is manually invoked
 *    on two versions of the same file to ensure that they are the same.
 *    Prints the the filename, size in bytes or sectors depending on the
 *    option flag, and the CRC.
 *
 *    Options:
 *        -t    Perform CRCK on a CP/M text file that is stored as a Unix
 *              file.  CP/M record delimters are substituted for Unix 
 *		newlines and the file is padded to a 128-byte sector
 *  		boundary with ^Z's.
 *
 *	  -u    Perform CRCK on a regular Unix file.  Takes whatever is there.
 *
 *	  -c    Perform CRCK on a CP/M ".COM" file.  Handled the same as
 *		the -u flag except sectors are reported rather than bytes.
 *
 *	  -i    Perform CRCK on a CP/M ".COM" file stored in ITS binary 
 *		format.  Ignores the first four bytes of the file.  Implies
 *  		the -c option.
 * 
 *	  If no option flag is chosen, the default option is -u.
 *    
 *    Wildcards may, of course, be used and the program may be used in 
 *    pipelines.
 *
 *
 *----------------------------------------------------------------------------
 *    Version 2.1  Cleanup by Glitch Works, LLC for modern C compilers,
 *                 http://www.glitchwrks.com/
 *
 *                 Added Makefile, patches against original code.
 *
 *    Version 2.0  mods by Ben Goldfarb (decvax!ucf-cs!goldfarb or
 *                                            Goldfarb.ucf-cs @ Rand-Relay)
 *
 *		   Added CP/M text and binary file modes and changed to 
 *		   use standard stream I/O.
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

#define VERSION 21		/*  Version 2.1				*/


#define SECSIZ	128		/*  CP/M sector size for fill  		*/
#define CTLZ 	'Z' - 0x40	/*  Fill character for text files	*/ 

int cflag = 0;			/*  CP/M binary file			*/
int tflag = 0;			/*  CP/M text file			*/
int uflag = 0;			/*  Unix file				*/
int iflag = 0;			/*  ITS Binary file			*/

void docrck(char *name);
unsigned short crck();		/*  CRC calculation function		*/


int main(argc, argv)
int argc;
char **argv;
{
	int err = 0;

	if (argc > 1 && (++argv)[0][0] == '-') {
		switch (argv[0][1]) {
		  case 't':
			++tflag;
			break;
		  case 'c': 
			++cflag;
			break;
		  case 'u':
			++uflag;
			break;
		  case 'i':
			++iflag; ++cflag;
			break;
		  default:
			goto usage;
		}
		--argc; ++argv;
	}
	else  /* for right now, "Unix mode" is the default */
		++uflag;

	if (isatty(2)) {  /*  print header if not in a pipe  */
		printf("CRCK program for Unix\n");
		printf("Version %1d.%1d\n\n", VERSION/10, VERSION%10);
		if (iflag) printf("ITS mode selected\n");
		if (tflag) printf("CP/M text mode selected\n");
		if (cflag) printf("CP/M \".COM\" mode selected\n");
		if (uflag) printf("Unix file mode selected\n");
		putchar('\n');
	}

	if (argc == 1 ) {  /*  we're just taking stdin from somewhere  */
		docrck("stdin");
		exit(0);
	}
	else {  /*  we're doing a named file or a list of same  */
		for (; --argc; ++argv) {
			err = 0;
			if (freopen(*argv, "r", stdin) == NULL) {
				perror(*argv);
				++err;
			}
			if (!err)
				docrck(*argv);
		}
		exit(0);
	}

usage:
	fprintf(stderr, "Usage: crck [-t|c|u|i] [filename ... ]\n");
	exit(-100);
}

void docrck(name) /* Calculate and print a "CRCK" for stdin stream. */
char *name;
{
	int icount;
	long nsec;
	unsigned short oldcrck;
	unsigned char crbuf[SECSIZ];
	unsigned char *bufend = crbuf + SECSIZ;

	register int c;
	register unsigned char *cp;

	oldcrck = 0;
	nsec = 0L;
	cp = crbuf;
	if (tflag) {
	    /*
   	     *  To calculate CRC for CP/M text files that are
	     *  presently stored in the form of Unix text files, 
	     *  an ASCII CR must be inserted in the buffer before 
             *  each '\n'.  Then the file must  be padded out to 
	     *  an even SECSIZ boundary with CTLZ's.
	     */
            while ((c = getchar()) != EOF) {
		if (((*cp++ = c) == '\n') && tflag) {  /*  need to add \r */
			*(cp - 1) = '\r';
			if (cp >= bufend) {  /* sector boundary */
				oldcrck = crck(crbuf, SECSIZ, oldcrck);
				++nsec;
				cp = crbuf;
			}
			*cp++ = '\n';
		}
		if (cp >= bufend) {  /* sector boundary */
			oldcrck = crck(crbuf, SECSIZ, oldcrck);
			++nsec;
			cp = crbuf;
		}
	    }

	    if (cp != crbuf) { /* EOF during partial sector  */
		register int fillcnt;

		/*  Fill from current pointer pos to end of buffer  */
		if (tflag)
			for (fillcnt = bufend - cp; fillcnt--; *cp++ = CTLZ)
					;

		/*  Need to update CRC for final sector. */
		oldcrck = crck(crbuf, cp-crbuf, oldcrck);
		++nsec;
	    }
	}
	else {  /* uflag or cflag  -- do a quick crck  */
	    unsigned char tbuf[BUFSIZ];
	    int count = 0;
	    
	    if (iflag) /* skip first four bytes of ITS binary file */
		for (icount = 0; 
		    (c = getchar()) != EOF && icount < 4;
		    icount++)
			;
	    while ((count = fread(tbuf, 1, BUFSIZ, stdin)) > 0) {
		oldcrck = crck(tbuf, count, oldcrck);
		if (!cflag)
			nsec += count;  /* here nsec is for BYTES  */
		else
			nsec += ((count + SECSIZ - 1) / SECSIZ);
	    }
	}

	printf("%-18s", name);
	printf(" (%7ld %-7s)   ==>   %04x\n", 
		nsec, uflag ? "bytes" : "sectors", oldcrck);
}

unsigned short
crck(crbuf, count, ldcrc)  /* uses algorithm of CRCK.COM prior to V5.0 */
register unsigned char *crbuf;
register int count;
register unsigned short ldcrc;
{
/*
 *  This routine copyright (c) 1983, Computer Development, Inc., 
 *  Beaverton, OR, USA.
 *  All rights reserved.
 */
	register unsigned short n;

	while (--count >= 0) {
		n = (ldcrc << 1);
		n = (n & 0xff00) | (( n + *crbuf++ ) & 0x00ff);
		if(ldcrc & 0x8000)
			n ^= 0xa097;
		ldcrc = n;
	}
	return(ldcrc);		
}
